crates/pop-fork/src/cache.rs

```diff
@@ -5,100 +5,217 @@
+				std::fs::create_dir_all(parent)?;
+			}
+			let url = path.display().to_string();
+
+			// Run migrations on a temporary async connection first
+			{
+				let mut conn = SyncConnectionWrapper::<SqliteConnection>::establish(&url).await?;
+				// Apply pragmas for better concurrency on file databases
+				// WAL mode: Persists to the database file itself
+				diesel::sql_query("PRAGMA journal_mode=WAL;").execute(&mut conn).await?;
+				// Busy timeout: For this migration connection
+				diesel::sql_query("PRAGMA busy_timeout=5000;").execute(&mut conn).await?;
+				let mut harness = AsyncMigrationHarness::new(conn);
+				harness.run_pending_migrations(MIGRATIONS)?;
+				let _ = harness.into_inner();
+			}
+
+			// Build the pool with connection customizer
+			let manager =
+				AsyncDieselConnectionManager::<SyncConnectionWrapper<SqliteConnection>>::new(url);
+			let pool = Pool::builder()
+				.max_size(MAX_POOL_CONNECTIONS)
+				.connection_customizer(Box::new(SqliteConnectionCustomizer))
+				.build(manager)
+				.await?;
+			Ok(Self { inner: StorageConn::Pool(pool) })
+		} else {
+			// Single in-memory connection
+			let mut conn = SyncConnectionWrapper::<SqliteConnection>::establish(":memory:").await?;
+			// Run migrations on this single connection
+			// Set busy timeout to reduce lock errors under contention
+			diesel::sql_query("PRAGMA busy_timeout=5000;").execute(&mut conn).await?;
+			let mut harness = AsyncMigrationHarness::new(conn);
+			harness.run_pending_migrations(MIGRATIONS)?;
+			let conn = harness.into_inner();
+			Ok(Self {
+				inner: StorageConn::Single(std::sync::Arc::new(tokio::sync::Mutex::new(conn))),
+			})
 		}
-
-		let url = format!("sqlite:{}?mode=rwc", path.display());
-		let pool = SqlitePoolOptions::new()
-			.max_connections(MAX_POOL_CONNECTIONS)
-			.connect(&url)
-			.await?;
-
-		// Create tables
-		sqlx::query(CREATE_TABLES_SQL).execute(&pool).await?;
-
-		Ok(Self { pool })
 	}
 
-	/// Open an in-memory cache (for testing).
-	#[cfg(test)]
+	/// Open an in-memory cache.
+	///
+	/// Creates a fresh in-memory SQLite database and runs all migrations
+	/// to set up the storage and blocks tables.
 	pub async fn in_memory() -> Result<Self, CacheError> {
-		let pool = SqlitePoolOptions::new()
-			.max_connections(MEMORY_POOL_CONNECTIONS)
-			.connect(SQLITE_MEMORY_URL)
-			.await?;
-
-		sqlx::query(CREATE_TABLES_SQL).execute(&pool).await?;
+		Self::open(None).await
+	}
 
-		Ok(Self { pool })
+	/// Get a database connection.
+	///
+	/// Handles acquiring the connection from either the pool or single mutex.
+	/// The connection is automatically returned to the pool or unlocks the mutex when dropped.
+	async fn get_conn(&self) -> Result<ConnectionGuard<'_>, DieselError> {
+		match &self.inner {
+			StorageConn::Pool(pool) => {
+				let conn = pool.get().await.map_err(|e| {
+					DieselError::DatabaseError(
+						DatabaseErrorKind::UnableToSendCommand,
+						Box::new(e.to_string()),
+					)
```

Comment:
Are we mapping to the correct error ?

I see what we have could make sense when a query has failed because we weren't able to acquire the connection.

But if we fail on a regular `let mut conn = self.get_conn().await?;` which we have a few, we will be using an odd error for it.

Maybe we should move from `DieselError` to `CacheError` 🤔 , what do you think @tsenovilla ?